### PR TITLE
feat: add smtp-client

### PR DESCRIPTION
**Pull request by tanguyantoine**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/smtp-client/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    constructor(options?: SMTPClientOptions);
+    connect(options?: ConnectOptions): Promise<void>;
+    authPlain(options?: AuthPlainOptions): Promise<void>;
+    authLogin(options?: AuthLoginOptions): Promise<void>;
+    close(options?: AuthLoginOptions): Promise<void>;
+    greet(options?: GreetOptions): Promise<void>;
+    mail(options?: MailOptions): Promise<void>;
+    rcpt(options?: RcptOptions): Promise<void>;
+    data(data: string | Buffer, options?: DataOptions): Promise<void>;
+    quit(options?: QuitOptions): Promise<void>;
```
Comment (martin-badin) [inline types/smtp-client/index.d.ts:18]: missing `helo`, `ehlo`, `hasExtension`, `getDataSizeLimit`, `getAuthMechanisms`, `parseEnhancedReplyCode`, `parseReplyText`, `noop`, `rset`, `secure`

https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L44
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L68
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L101
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L109
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L123
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L142
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L151
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L215
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L236
https://github.com/xpepermint/smtp-client/blob/master/src/index.js#L313

File: types/smtp-client/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+export interface ConnectOptions {
+    timeout?: number;
+}
+export interface AuthPlainOptions {
+    username?: string;
+    password?: string;
+}
+export interface AuthLoginOptions {
+    username?: string;
+    password?: string;
```
Comment (martin-badin) [inline types/smtp-client/index.d.ts:34]: missing `timeout`
```suggestion
    password?: string;
    timeout?: number;
```

File: types/smtp-client/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+export interface SMTPClientOptions {
+    host?: string;
+    port?: number;
+}
+export interface ConnectOptions {
+    timeout?: number;
+}
+export interface AuthPlainOptions {
+    username?: string;
+    password?: string;
```
Comment (martin-badin) [inline types/smtp-client/index.d.ts:30]: missing `timeout`
```suggestion
    password?: string;
    timeout?: number;
```

**Review by martin-badin — CHANGES_REQUESTED** [review]

File: types/smtp-client/index.d.ts (hunk 1 above)
Comment (tanguyantoine) [reply types/smtp-client/index.d.ts]: done

File: types/smtp-client/index.d.ts [hunk 4]
```diff
@@ -7,52 +7,71 @@
...
+export interface EhloOptions extends TimeoutOptions {
+    hostname?: string;
 }
 
 export interface SMTPClientOptions {
     host?: string;
     port?: number;
 }
+
 export interface ConnectOptions {
```
Comment (martin-badin) [inline types/smtp-client/index.d.ts:47]: Please extends this from `TimeoutOptions`

**Review by martin-badin — CHANGES_REQUESTED** [review]

File: types/smtp-client/index.d.ts (hunk 4 above)
Comment (tanguyantoine) [reply types/smtp-client/index.d.ts]: Good catch. I removed it in favor of timeout option

File: types/smtp-client/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    host?: string;
+    port?: number;
+}
+
+export interface AuthPlainOptions extends TimeoutOptions {
+    username?: string;
+    password?: string;
+}
+
+export interface AuthLoginOptions extends TimeoutOptions {
```
Comment (martin-badin) [inline types/smtp-client/index.d.ts:52]: Maybe you can merge `AuthLoginOptions` and `AuthPlainOptions` to one interface with name `AuthOptions`

Comment (tanguyantoine) [reply types/smtp-client/index.d.ts:52]: Can you provide a PR on top of mine for this kind of changes?

File: types/smtp-client/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    utf8?: boolean;
+}
+
+export interface RcptOptions extends TimeoutOptions {
+    to?: string;
+}
+
+export interface DataOptions extends TimeoutOptions {
+    sourceSize?: number;
+}
```
Comment (martin-badin) [inline types/smtp-client/index.d.ts:72]:
```suggestion
// Type definitions for smtp-client 0.3
// Project: https://github.com/xpepermint/smtp-client#readme
// Definitions by: TANGUY Antoine <https://github.com/tanguyantoine>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

/// <reference types="node" />

export class SMTPClient {
    constructor(options?: SMTPClientOptions);
    authLogin(options?: AuthOptions): Promise<void>;
    authPlain(options?: AuthOptions): Promise<void>;
    close(options?: TimeoutOptions): Promise<void>;
    connect(options?: TimeoutOptions): Promise<void>;
    data(data: string | Buffer, options?: DataOptions): Promise<void>;
    mail(options?: MailOptions): Promise<void>;
    quit(options?: TimeoutOptions): Promise<void>;
    rcpt(options?: RcptOptions): Promise<void>;
    greet(options?: GreetOptions): Promise<void>;
    helo(options?: HeloOptions): Promise<void>;
    ehlo(options?: EhloOptions): Promise<void>;
    hasExtension(extensions: string[]): boolean;
    getDataSizeLimit(): number;
    getAuthMechanisms(): string[];
    parseEnhancedReplyCode(line: string): string | null;
    parseReplyText(line: string): string;
    noop(options?: TimeoutOptions): Promise<void>;
    secure(options?: TimeoutOptions): Promise<void>;
}

export interface TimeoutOptions {
    timeout?: number;
}

export interface HeloOptions extends TimeoutOptions {
    hostname?: string;
}

export interface EhloOptions extends TimeoutOptions {
    hostname?: string;
}

export interface SMTPClientOptions {
    host?: string;
    port?: number;
}

export interface AuthOptions extends TimeoutOptions {
    username?: string;
    password?: string;
}

export interface GreetOptions extends TimeoutOptions {
    hostname?: string;
}

export interface MailOptions extends TimeoutOptions {
    from?: string;
    utf8?: boolean;
}

export interface RcptOptions extends TimeoutOptions {
    to?: string;
}

export interface DataOptions extends TimeoutOptions {
    sourceSize?: number;
}
```

**Comment by tanguyantoine** [comment]
I'll create a PR within the project itself

**Pull request closed without merging**